include/watchdog.h

```diff
@@ -1,116 +1,219 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
+/**
+ * @brief Watchdog Interface
+ * @defgroup watchdog_interface Watchdog Interface
+ * @ingroup timer_counter_interfaces
+ * @{
+ */
+
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_OPT_PAUSE_IN_SLEEP	BIT(0)
 
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
-};
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_OPT_PAUSE_HALTED_BY_DBG	BIT(1)
 
 /**
- * WDT clock cycles for timeout type.
- */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+ * @brief Run callback function on watchdog timeout. If any of reset flags is
+ *	  set this callback is called before reset.
+ */
+#define WDT_FLAG_RUN_CALLBACK	BIT(0)
+
+/**
+ * @brief Watchdog reset behavior.
+ *
+ * If none of the reset flags is specified there will be no reset after timeout.
+ */
+/** CPU core reset */
+#define WDT_FLAG_RESET_CPU	BIT(1)
+/** CPU core and peripherals reset */
+#define WDT_FLAG_RESET_CPU_PER	BIT(2)
+/** Global system reset */
+#define WDT_FLAG_RESET_SOC	(BIT(2) | BIT(1))
 
+/**
+ * @brief Watchdog reset flag bit field mask.
+ */
+#define WDT_FLAG_RESET_MASK	(0x3 << 1)
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs reloading in specified time window, otherwise
+ * the watchdog will trigger. If watchdog instance does not support window
+ * timeouts then min value must be equal to 0.
+ *
+ * @param min Lower limit of watchdog reload timeout in microseconds.
+ * @param max Upper limit of watchdog reload timeout in microseconds.
```

Comment:
I have come back to microseconds and 32 bit values. The question is if there is any need to return actual set timeout values. For example the resolution of watchdog timer is 60 us and super supply 100 us. I think there is no need because users usually estimate timeout values and this estimation is rather coarse. Maybe we should only add information to the documentation that these values are always rounded up and they may vary depending on platform?

Reply:
SYS_LOG_INFO with real timeout is great idea - as it is indeed debug related functionality. So I would be up to the will of MCU ports maintainers.

Reply:
Is adding a `@note` to the documentation of `wdt_setup` that it is welcomed to print actual timeouts using `SYS_LOG_INFO` an abuse?